torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -109,52 +110,125 @@ std::shared_ptr<FutureMessage> sendPythonRemoteCall(
-    const std::shared_ptr<FutureMessage>& fm,
+c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
+    const std::shared_ptr<FutureMessage>& futureResponseMessage,
     bool hasValue) {
-  auto fv = std::make_shared<FutureIValue>();
-
-  fm->addCallback([fv, hasValue](const FutureMessage& fm) {
-    if (fm.hasError()) {
-      fv->setError(*fm.error());
-    } else if (hasValue) {
-      // Don't need to acquire GIL here, as toPyObj and toIValue acquires GIL
-      // when creating/copying the py::object
-      fv->markCompleted(
-          jit::toIValue(toPyObj(fm.constValue()), PyObjectType::get()));
-    } else {
-      fv->markCompleted(IValue());
-    }
-  });
-
-  return fv;
+  if (hasValue) {
+    // NB: The custom deleter is necessary because the FutureIValue object
+    // holds a py::object and it would require GIL to delete.
+    c10::intrusive_ptr<JitFuture> jitFuture =
+        c10::make_intrusive<JitFuture>(PyObjectType::get());
+
+    futureResponseMessage->addCallback(
+        [jitFuture](const FutureMessage& futureResponseMessage) {
+          // Don't need to acquire GIL here, as toPyObj acquires GIL
+          // when creating the py::object
+          if (futureResponseMessage.hasError()) {
+            jitFuture->setError(futureResponseMessage.error()->what());
+          } else {
+            jitFuture->markCompleted(jit::toIValue(
+                toPyObj(futureResponseMessage.constValue()),
+                PyObjectType::get()));
+          }
+        });
+
+    return jitFuture;
+  } else {
+    c10::intrusive_ptr<JitFuture> jitFuture =
+        c10::make_intrusive<JitFuture>(NoneType::get());
+
+    futureResponseMessage->addCallback(
+        [jitFuture](const FutureMessage& futureResponseMessage) {
+          if (futureResponseMessage.hasError()) {
+            jitFuture->setError(c10::ivalue::Future::FutureError(
+                futureResponseMessage.error()->what()));
+          } else {
+            jitFuture->markCompleted(IValue());
+          }
+        });
+
+    return jitFuture;
+  }
 }
 
-std::shared_ptr<FutureIValue> pyRpcBuiltin(
+c10::intrusive_ptr<JitFuture> pyRpcBuiltin(
     const WorkerInfo& dst,
     const std::string& opName,
+    const float rpcTimeoutSeconds,
     const py::args& args,
-    const py::kwargs& kwargs,
-    const float rpcTimeoutSeconds) {
+    const py::kwargs& kwargs) {
+  DCHECK(PyGILState_Check());
   Stack stack;
   auto op = matchBuiltinOp(opName, args, kwargs, stack);
   // Release GIL since args and kwargs processing is done.
   py::gil_scoped_release release;
   auto scriptCall = std::make_unique<ScriptCall>(op, std::move(stack));
   auto agent = RpcAgent::getCurrentRpcAgent();
-  return toFutureIValue(sendMessageWithAutograd(
+  return wrapFutureMessageInJitFuture(sendMessageWithAutograd(
       *agent,
       dst,
       std::move(*scriptCall).toMessage(),
       false,
       rpcTimeoutSeconds));
 }
 
+c10::intrusive_ptr<JitFuture> pyRpcPythonUdf(
```

Comment:
There is no change in the following two functions, instead they are just moved here, right?

Reply:
Right, I am moving. Since @wanchaol 's [comment](https://github.com/pytorch/pytorch/pull/35154#discussion_r396673033) above have some confusions about the interfaces exposed by "python_functinos.h" and "torchscript_functions.h", so I want to clearly list them out here.